README.md
# Evaluation Examples Are Not Equally Informative: How Should That Change NLP Leaderboards?

This repository accompanies our ACL 2021 paper [Evaluation Examples Are Not Equally Informative: How Should That Change NLP Leaderboards?]()

The code here includes what is needed to reproduce those results as well additional work that did not make it into the publication.
Eventually, our goal is that this code provides a re-usable implementation of IRT models geared towards use in NLP leaderboards.

NOTE: There is one set of results that will not be publicly reproducible, namely those that require instance level scores on the SQuAD test set.
To maintain the integrity of the benchmark task, we cannot release this data.

## Running Our Code

We support (at least) two use cases with this code:

1. (Part of a) Full reproduction of our experimental results from scratch
2. A way to quickly run our visualizations without rerunning everything from scratch

For (1), continue reading these instructions, for (2) skip to the bottom.

Our code is broken into two repositories: (1) this repository and (2) the [repository with the LaTeX source code of our paper](https://github.com/entilzha/publications). Roughly, this code encompasses most of the heavy lifting (e.g., running models), but is not mainly responsible for generating the tables and plots in the paper. Instead, this code produces data in a folder called `data` which is then used by the `figures.py` file in the paper source code to generate all relevant figures and statistics.

## Installation

### Installation (required)

- Install Anaconda python
- Create an environment like `conda create -n leaderboard python=3.8`
- Install [Python Poetry](python-poetry.org/) then run `conda activate leaderboard` followed by `poetry install`
- Install Altair dependencies with `conda install -c conda-forge vega-cli vega-lite-cli`
- Install Vowpal Wabbit https://github.com/VowpalWabbit/vowpal_wabbit/

When running project code, be sure to source the conda environment with `conda activate leaderboard` first.

### Installation (frontend, optional)

Separate from our ACL paper, for a course project we developed a proof of concept leaderboard website at [leaderboard.pedro.ai](https://leaderboard.pedro.ai). These instructions explain how to install the dependencies for this. Note: this is different from the UI discussed in the ACL paper.

The frontend is written in typescript + react + d3/vega + materialui.
To run the frontend

1. Install `yarn`
2. `cd frontend`
3. `yarn install`
4. So that the backend and frontend can talk to each other, we use [Caddy](https://caddyserver.com).

## Data

All the data used in our experiments can be downloaded or generated using the instructions below.
The one exception to this are instance-level test scores of SQuAD models, which we cannot release.

### Download Data

If you only need to run our visualizations or the [paper generation code](https://github.com/entilzha/publications), follow these instructions to download a copy of our data.

There are three versions of our data, of decreasing sizes, download one that matches your use case.

- Full data, decompressed 20-25GB: `wget https://obj.umiacs.umd.edu/acl2021-leaderboard/leaderboard-data-all.tar.gz`
- Data minus linear models, decompressed 6GB: `wget https://obj.umiacs.umd.edu/acl2021-leaderboard/leaderboard-data-minus-linear.tar.gz`
- Only IRT related models/data, decompressed 1GB: `wget https://obj.umiacs.umd.edu/acl2021-leaderboard/leaderboard-data-only-irt.tar.gz`

If this repository is cloned to `irt-leaderboard`, you should save the `tar.gz` file to `irt-leaderboard.tar.gz`.
Then, run `tar xzvf irt-leaderboard.tar.gz` which will decompress the tar into `data/`.

If data already exists in `data/`, then you should decompress in a different directory and use a command like `rclone copy other-dir/data irt-leaderboard/data` to copy files (this will override any duplicate files).

### Creating Data from Scratch

To reproduce our results "from scratch," follow these instructions.
When complete, this is equivalent to having followed the "Download Data" section.

Broadly, this is broken into two steps:

1. Downloading SQuAD predictions from [codalab.org](https://codalab.org)
2. Running our data pipeline

#### Downloading SQuAD predictions

Before anything, you should download a copy of the [SQuAD development set](https://rajpurkar.github.io/SQuAD-explorer/dataset/dev-v2.0.json) and save it to `data/squad/datasets/dev-v2.0.json`.
You'll also need to download the training sets for both datasets and save it in a similar location (see `config.toml` for exact location).

Download the `out-v2.0.json` and `out-v1.0.json` files from https://github.com/rajpurkar/SQuAD-explorer and place them in `data`.
We used the---at the time---latest version of this file, which has likely changed since we ran our experiments.
Specifically, our version has an `updated` timestamp of `1601497446.0011725`.

Some scripts that access codalab are modified versions of the scripts found here: https://github.com/codalab/codalab-worksheets/blob/master/scripts/competitiond.py

To download codalab data, you'll need a codalab.org account.
For convenience, you can set your username/password with environment variables `CODALAB_USERNAME` and `CODALAB_PASSWORD`.

This file is used by `leaderboard/codalab.py` to determine which codalab bundles (predictions) to download.
Specifically, you should run this file to download each submission file to `data/squad/submissions/` and corresponding metadata to `data/squad/submission_metadata.json` + `data/squad/skipped_submissions.json`:

- `leaderboard cl download`

Since these are only the predictions, next you will need to score them like this (saves to `data/squad/submission_predictions.json`):

- `leaderboard cl score-submissions`

#### Creating the Submission Database

For historical reasons (web app), our downstream pipelines read a sqlite database holding all the SQuAD model submissions.

Once you have all the files from previous steps, run this to create the database (`data/explorer.sqlite3`):

- `leaderboard squad init-db --skip-tests`

If your database is in a corrupt state before running the command, you may need to delete the bad database first.

As a sanity check, your database should look like this:

```
❯ litecli data/explorer.sqlite3                                                                                                                   (leaderboard)
Version: 1.6.0
Mail: https://groups.google.com/forum/#!forum/litecli-users
GitHub: https://github.com/dbcli/litecli
data/explorer.sqlite3> .tables
+-------------------+
| name              |
+-------------------+
| examples          |
| predictions       |
| statistical_tests |
| submissions       |
+-------------------+
Time: 0.010s
data/explorer.sqlite3> select count() from examples;
+---------+
| count() |
+---------+
| 11873   |
+---------+
1 row in set
Time: 0.006s
data/explorer.sqlite3> select count() from predictions;
+---------+
| count() |
+---------+
| 1911553 |
+---------+
1 row in set
Time: 0.014s
data/explorer.sqlite3> select count() from statistical_tests;
+---------+
| count() |
+---------+
| 0       |
+---------+
1 row in set
Time: 0.005s
data/explorer.sqlite3> select count() from submissions;
+---------+
| count() |
+---------+
| 161     |
+---------+
1 row in set
Time: 0.006s
```

#### Running Data Pipeline

Our data pipeline is handled by [Luigi](https://github.com/spotify/luigi), a make-like python library.
Like make, it functions by listing a set of [tasks](https://github.com/EntilZha/irt-leaderboard/blob/master/leaderboard/tasks.py) that each have:

- pre-requisitives (`def requires()`)
- run definitions (`def run()`)
- outputs (`def output()`)

The pipeline requires that a few files exist before it will run (dependencies external to the pipeline, hence they inherit from `luigi.ExternalTask`).

- `class Database`: `data/explorer.sqlite3`
- `class SquadV2`: SQuAD train/dev sets, saved to `data/squad/datasets/train-v1.1.json`, `data/squad/datasets/train-v2.0.json`, and `data/squad/datasets/dev-v2.0.json`
- `class SugawaraFeatures`: SQuAD easy and hard subsets from https://github.com/Alab-NII/mrc-heuristics/tree/master/subsets, saved to `data/squad/datasets/squad-easy-subset.json` and `data/squad/datasets/squad-hard-subset.json`
- `class SquadOutV2`: Previously mentioned `out-v2.0.json` file, saved to `data/squad/out-v2.0.json`
- `class LeaderboardSubmissionPredictions`: Downloaded from codalab, corresponds to `data/squad/submission_predictions.json`

Once these are present, you can run all the experiments using these commands:

- `luigi --module leaderboard.tasks --workers 1 --local-scheduler AllTopicModels`
- `luigi --module leaderboard.tasks --workers 1 --local-scheduler AllIRTEvaluations`
- `luigi --module leaderboard.tasks --workers 1 --local-scheduler AllVWModels`

To dry run these commands, simply change this argument `--workers 0`.
This is helpful to check if all the pre-requisite external data has been downloaded or to check if all the pre-computed data has been downloaded correctly (in which case it should report there are no tasks to run).

#### Running Components of Pipeline

Much of our code is written so that it can either be called as part of the pipeline, or with the CLI `leaderboard` which you can see more about in `leaderboard/main.py` or calling `leaderboard --help`

## Simulations

In our paper, we run several simulations to comparing ranking stability and annotation sampling strategies.
These are not run by the pipeline, but some depend on having run certain pipeline elements first.
We include instructions for running both simulations, but be aware that only the simulations that do not require SQuAD test data will be reproducible to those without that data.
Namely, the main thing that is reproducible is comparing the rankings of two mutually exclusive samples of the development set.

NOTE: Most of these simulations are relatively CPU intensive and take some time to run.

### Ranking Stability Simulation

These instructions reproduce the experiments in Figure 4 that compare the ranking stability of accuracy rankings versus IRT rankings:

![Ranking Stability Comparison](images/stability_simulation_corr.png)

Due to the compute intensive nature of this experiment, we ran these experiments in the CLIP lab's CPU compute cluster which uses slurm.
If you do not use slurm, it is still possible to run our experiments, but it will take some adaptation.
Please create a github issue if you have any trouble with this and need pointers.

In the following instructions, `/fs/clip-quiz/entilzha/ranking_stability` refers to the path storing the scripts that slurm eventually runs.
These scripts call `bin/slurm-run-rank-stability.sh` with the appropriate arguments.
An example of what step 1 generates is:

```
# Contents of run.sh
sbatch bin/slurm-run-rank-stability.sh 25 0 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 25 1 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 25 2 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 25 3 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 25 4 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 25 5 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 25 6 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 25 7 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 25 8 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 25 9 /fs/clip-quiz/entilzha/ranking_stability

sbatch bin/slurm-run-rank-stability.sh 50 0 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 50 1 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 50 2 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 50 3 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 50 4 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 50 5 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 50 6 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 50 7 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 50 8 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 50 9 /fs/clip-quiz/entilzha/ranking_stability

sbatch bin/slurm-run-rank-stability.sh 75 0 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 75 1 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 75 2 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 75 3 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 75 4 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 75 5 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 75 6 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 75 7 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 75 8 /fs/clip-quiz/entilzha/ranking_stability
sbatch bin/slurm-run-rank-stability.sh 75 9 /fs/clip-quiz/entilzha/ranking_stability
...

# Contents of config.json
{"fold": "dev", "irt_type": "3PL"}
```

In our experiments, we ran `leaderboard rank_stability configure` twice.
The first created simulations of size 25, 50, and 75 to cover lower sample sizes well.
The second created more spaced out simulations from 100 to 5000 in steps of 100.

1. Configure the simulation: `leaderboard rank_stability configure dev 3PL /fs/clip-quiz/entilzha/ranking_stability`
2. Run the simulation: `sbatch /fs/clip-quiz/entilzha/ranking_stability/run.sh`
3. Create the plot `leaderboard plot --plot stability`

### Cold Start Reliability Simulation

To run the cold start reliability simulation, run `python leaderboard/cat_sampling_stability.py`.

## Supplying Data to Publications Repository

As previously mentioned, we break our source code into two components: (1) the experimental code and plotting code heavily integrated into experimental code and (2) LaTeX source code and the rest of our plotting code.

Once you have run the experiments (or downloaded the data), the best way to link the two repositories up is by creating a symbolic link.
Assuming the two repositories are at `~/code/publications` and `~/code/irt-leaderboard`, run:

```
$ cd publications
$ mkdir 2021_acl_leaderboard/auto_data
$ ln -s ~/code/irt-leaderboard/data ~/code/publications/auto_data/data
```

## Quick Start: Running Visualizations

To run our visualization demo:

1. TODO: Download the data
2. Run `conda activate leaderboard`
3. Run `streamlit run leaderboard/visualize.py`

### Internal to UMD

To save data to the public bucket, use this command:

```bash
$ # To full sync
$ rclone sync  data/ obj:acl2021-leaderboard/data/
$ # To ignore the heavy VW files
$ rclone --checksum --exclude-from rclone-exclude.txt sync  data/ obj:acl2021-leaderboard/data/
```

Where `obj` refers to an rclone key, with a configuration like below, with keys set appropriately:

```
[obj]
type = s3
provider = Other
env_auth = false
access_key_id = CHANGEME
secret_access_key = CHANGEME
endpoint = http://obj.umiacs.umd.edu
```

Similar commands can also be used to sync in the opposite direction.

To generate the tarfiles:

- all the data: `tar czvf` the data
- minus linear: delete `data/linear` then `tar czvf`
- only irt: delete `data/linear`, `data/simulations`, `data/mlqa`, and `data/topic-model` then `tar czvf`

# Additional Instructions

Not all our experiments made it into the paper.
Instructions below pertain to those

## Running the Frontend/Backend for leaderboard.pedro.ai

For a course project (before working on our paper), we prototyped the leaderboard interface at leaderboard.pedro.ai (different from the demo in our paper).
To run this, follow the instructions below
To run the frontend and backend, in separate terminals run:

- Reverse proxy: `caddy start`
- Frontend: `cd frontend` then `yarn start`
- Backend: `poetry run uvicorn --reload leaderboard.www.app:app`
- Yarn will open `localhost:3000`, but you need to instead navigate to `localhost:2000` which is the address the caddy proxy uses

## Rerun Squad with MLQA Data

Although we used the scores of test examples, we cannot release this.
An alternative to this challenge is to use MLQA which is in the same format as SQuAD.
To do this, we need to replicate a similar process that the Squad organizers use to rerun the submitted models on an alternative test set; this section documents this process.
This said, this did not make it into our paper since many models we tried did not successfully run, got stuck while running, or we otherwise had other technical problems.
We leave these instructions/this idea here in case an enterprising undergraduate/graduate student is interested in making this work.

Squad Scripts

- Leaderboard: https://github.com/rajpurkar/SQuAD-explorer
- Competition Update Script: https://github.com/codalab/codalab-worksheets/blob/master/scripts/competitiond.py

Bundle IDs for MLQA data

- Worksheet: https://worksheets.codalab.org/worksheets/0xf5b78f83e1a74d4ea8226b93d4d1198e

CHANGELOG.md
# CHANGELOG

We do not have regular releases, so there are no changes to log version to version.

leaderboard/tasks.py
"""Copyright (c) Facebook, Inc. and its affiliates."""
import random
import subprocess
from pathlib import Path
from typing import Optional

import luigi
from functional import pseq
from luigi.local_target import LocalTarget
from pedroai.io import read_json, read_jsonlines, write_json, write_jsonlines
from sklearn.model_selection import train_test_split

from leaderboard import codalab, power
from leaderboard.config import DATA_ROOT, conf
from leaderboard.data import create_leaderboard_splits, squad_pred_scores_to_jsonlines
from leaderboard.irt.evaluate import evaluate_irt_model
from leaderboard.irt.model_svi import SVISquadIrt
from leaderboard.linear import eval_vw, to_vw
from leaderboard.log import get_logger
from leaderboard.stats import Sampling, StatTest, run_parallel_test
from leaderboard.topics import TopicModel
from leaderboard.www.database import export_submissions

log = get_logger(__name__)

IRT_FAMILIES = ["pyro", "stan", "no_irt"]
IRT_TYPES = ["1PL", "2PL", "3PL", "no_irt_pl"]
FEATURE_SETS = list(conf["vw"].keys())


def generate_vw_dir_path(*, irt_family: str, irt_type: str, feature_set: Optional[str]):
    base_path = Path("data/linear") / irt_family / irt_type
    if feature_set is None:
        return base_path
    else:
        return base_path / feature_set


def shell(command: str):
    log.info("SHELL: %s", command)
    subprocess.run(command, shell=True, check=True)


class Database(luigi.ExternalTask):
    def output(self):
        return luigi.LocalTarget(conf["db"])


class LeaderboardJsonlines(luigi.Task):
    fold = luigi.Parameter()

    def requires(self):
        return Database()

    def run(self):
        if self.fold == "dev":
            submissions = export_submissions()
        elif self.fold == "test":
            submissions = squad_pred_scores_to_jsonlines("test")
        else:
            raise ValueError("Invalid fold")
        write_jsonlines(conf["squad"]["leaderboard"][self.fold], submissions)

    def output(self):
        return luigi.LocalTarget(conf["squad"]["leaderboard"][self.fold])


class SquadV2(luigi.ExternalTask):
    def output(self):
        return [
            luigi.LocalTarget(DATA_ROOT / conf["squad"]["train_v1"]),
            luigi.LocalTarget(DATA_ROOT / conf["squad"]["train_v2"]),
            luigi.LocalTarget(DATA_ROOT / conf["squad"]["dev_v2"]),
        ]


class SugawaraFeatures(luigi.ExternalTask):
    def output(self):
        return [
            luigi.LocalTarget(DATA_ROOT / "data" / "squad" / "datasets" / "squad-easy-subset.json"),
            luigi.LocalTarget(DATA_ROOT / "data" / "squad" / "datasets" / "squad-hard-subset.json"),
        ]


class SquadOutV2(luigi.ExternalTask):
    def output(self):
        return luigi.LocalTarget(DATA_ROOT / conf["squad"]["out_v2"])


class PairedSquadDevTest(luigi.Task):
    def requires(self):
        yield SquadOutV2()

    def run(self):
        codalab.pair_squad_dev_test()

    def output(self):
        return luigi.LocalTarget(DATA_ROOT / conf["squad"]["dev_to_test"])


class LeaderboardSubmissionPredictions(luigi.ExternalTask):
    fold = luigi.ChoiceParameter(choices=["dev", "test"])

    def output(self):
        return luigi.LocalTarget(DATA_ROOT / conf["squad"]["submission_predictions"][self.fold])


class SquadTopics(luigi.ExternalTask):
    def output(self):
        return luigi.LocalTarget(DATA_ROOT / "data" / "squad-topics.txt")


class LeaderboardTrainTestSplit(luigi.Task):
    train_size = luigi.FloatParameter(0.9, significant=False)
    seed = luigi.IntParameter(42, significant=False)
    fold = luigi.Parameter()

    def requires(self):
        yield LeaderboardJsonlines(fold=self.fold)

    def run(self):
        create_leaderboard_splits(self.fold, seed=self.seed, train_size=self.train_size)

    def output(self):
        return luigi.LocalTarget(conf["squad"]["leaderboard_splits"][self.fold])


IRT_EVALUATION_TYPES = ["full", "heldout", "subject_heldout"]


class SplitDevSubjects(luigi.Task):
    """
    Split dev subjects for the sampling experiment
    """

    seed = luigi.IntParameter(42, significant=False)

    def requires(self):
        yield LeaderboardJsonlines(fold="dev")

    def run(self):
        random.seed(self.seed)
        subjects = read_jsonlines(conf["squad"]["leaderboard"]["dev"])
        train, heldout = train_test_split(subjects, train_size=0.8)
        train_ids = [s["submission_id"] for s in train]
        heldout_ids = [s["submission_id"] for s in heldout]
        out = {"train": train_ids, "heldout": heldout_ids}
        write_json(DATA_ROOT / conf["stability"]["dev_subject_splits"], out)

    def output(self):
        return luigi.LocalTarget(DATA_ROOT / conf["stability"]["dev_subject_splits"])


class SamplingIRTModel(luigi.Task):
    device = luigi.Parameter(default="cpu", significant=False)

    def requires(self):
        yield SplitDevSubjects()

    def run(self):
        model = SVISquadIrt(
            data_path=conf["squad"]["leaderboard"]["dev"],
            model="2PL",
            evaluation="subject_heldout",
        )
        model.train(device=self.device)
        model.save(
            DATA_ROOT
            / conf["irt"]["squad"]["dev"]["pyro"]["2PL"]["subject_heldout"]
            / "parameters.json"
        )

    def output(self):
        return luigi.LocalTarget(
            DATA_ROOT
            / conf["irt"]["squad"]["dev"]["pyro"]["2PL"]["subject_heldout"]
            / "parameters.json"
        )


class PyroTrainIRT(luigi.Task):
    irt_type = luigi.ChoiceParameter(choices=IRT_TYPES)
    evaluation = luigi.ChoiceParameter(choices=IRT_EVALUATION_TYPES)
    device = luigi.Parameter(default="cpu", significant=False)
    fold = luigi.ChoiceParameter(choices=["dev", "test"])

    def requires(self):
        yield LeaderboardJsonlines(fold=self.fold)
        yield LeaderboardTrainTestSplit(fold=self.fold)

    def run(self):
        model = SVISquadIrt(
            data_path=conf["squad"]["leaderboard"][self.fold],
            model=self.irt_type,
            evaluation=self.evaluation,
        )
        model.train(device=self.device)
        model.save(
            DATA_ROOT
            / conf["irt"]["squad"][self.fold]["pyro"][self.irt_type][self.evaluation]
            / "parameters.json"
        )

    def output(self):
        return luigi.LocalTarget(
            DATA_ROOT
            / conf["irt"]["squad"][self.fold]["pyro"][self.irt_type][self.evaluation]
            / "parameters.json"
        )


class EvaluateIRT(luigi.Task):
    irt_type = luigi.ChoiceParameter(choices=IRT_TYPES)
    evaluation = luigi.ChoiceParameter(choices=IRT_EVALUATION_TYPES)
    device = luigi.Parameter(default="cpu", significant=False)
    fold = luigi.ChoiceParameter(choices=["dev", "test"])

    def requires(self):
        yield PyroTrainIRT(
            evaluation=self.evaluation, irt_type=self.irt_type, device=self.device, fold=self.fold,
        )

    def run(self):
        evaluate_irt_model(
            evaluation=self.evaluation,
            model_type=self.irt_type,
            model_family="pyro",
            fold=self.fold,
        )

    def output(self):
        base_dir = (
            DATA_ROOT / conf["irt"]["squad"][self.fold]["pyro"][self.irt_type][self.evaluation]
        )
        return [
            luigi.LocalTarget(base_dir / "report.json"),
            luigi.LocalTarget(base_dir / "roc.pdf"),
            luigi.LocalTarget(base_dir / "precision_recall.pdf"),
        ]


class AllIRTEvaluations(luigi.WrapperTask):
    device = luigi.Parameter(default="cpu", significant=False)

    def requires(self):
        folds = ["dev"]
        if Path(conf["squad"]["submission_predictions"]["test"]).exists():
            folds.append("test")

        for fold in folds:
            for irt_type in IRT_TYPES:
                for evaluation in IRT_EVALUATION_TYPES:
                    if "no_irt" in irt_type or "no_irt" in evaluation:
                        continue

                    # Only need a trained model for the CAT sampling stability
                    # experiment
                    if evaluation == "subject_heldout":
                        if irt_type == "2PL" and fold == "dev":
                            yield PyroTrainIRT(
                                evaluation=evaluation,
                                irt_type=irt_type,
                                device=self.device,
                                fold=fold,
                            )
                    else:
                        yield EvaluateIRT(
                            evaluation=evaluation, irt_type=irt_type, device=self.device, fold=fold,
                        )


class VWTrainTestData(luigi.Task):
    irt_family = luigi.ChoiceParameter(choices=IRT_FAMILIES)
    irt_type = luigi.ChoiceParameter(choices=IRT_TYPES)
    feature_set = luigi.ChoiceParameter(choices=FEATURE_SETS)

    def requires(self):
        if self.irt_family == "pyro":
            # Use the best pyro model
            yield PyroTrainIRT(irt_type=self.irt_type, evaluation="full")
        elif self.irt_family == "stan":
            raise NotImplementedError
        elif self.irt_family == "no_irt":
            pass
        else:
            raise ValueError(f"Invalid model family: {self.irt_family}")
        yield LeaderboardTrainTestSplit(fold="dev")
        # yield SquadTopics()
        yield SugawaraFeatures()
        yield SquadV2()
        yield LeaderboardSubmissionPredictions(fold="dev")

    def run(self):
        out_dir = generate_vw_dir_path(
            irt_family=self.irt_family, irt_type=self.irt_type, feature_set=self.feature_set,
        )
        to_vw(
            irt_family=self.irt_family,
            irt_type=self.irt_type,
            feature_set=self.feature_set,
            out_dir=out_dir,
        )

    def output(self):
        out_dir = generate_vw_dir_path(
            irt_family=self.irt_family, irt_type=self.irt_type, feature_set=self.feature_set,
        )
        return (
            LocalTarget(out_dir / "train.vw.txt"),
            LocalTarget(out_dir / "test.vw.txt"),
        )


class VWHyperTune(luigi.Task):
    irt_family = luigi.ChoiceParameter(choices=IRT_FAMILIES)
    irt_type = luigi.ChoiceParameter(choices=IRT_TYPES)
    feature_set = luigi.ChoiceParameter(choices=FEATURE_SETS)

    def requires(self):
        return VWTrainTestData(
            irt_family=self.irt_family, irt_type=self.irt_type, feature_set=self.feature_set,
        )

    def run(self):
        # pylint: disable=line-too-long
        out_dir = generate_vw_dir_path(
            irt_family=self.irt_family, irt_type=self.irt_type, feature_set=self.feature_set,
        )
        args = [
            "python",
            "scripts/vw-hyperopt.py",
            "--max_evals",
            "20",
            "--train",
            str(out_dir / "train.vw.txt"),
            "--holdout",
            str(out_dir / "test.vw.txt"),
            "--best_vw_command_file",
            str(out_dir / "best-vw-command.json"),
            "--vw_space",
            "'--learning_rate=.001..10~L --passes=5 --l2=1e-8..1e-1~L0 -b=20,21,22,23 --link=logistic --loss_function=logistic'",
            "--out_dir",
            str(out_dir),
            "--plot",
        ]
        interactions = conf["vw"][self.feature_set].get("interactions")
        if interactions is not None:
            args.append("--additional_cmd")
            interaction_args = []
            for feature_interaction in interactions:
                interaction_args.append("--interactions")
                interaction_args.append(str(feature_interaction))
            additional_cmd = " ".join(interaction_args)
            args.append(f"'{additional_cmd}'")
        shell(" ".join(args))

    def output(self):
        out_dir = generate_vw_dir_path(
            irt_family=self.irt_family, irt_type=self.irt_type, feature_set=self.feature_set,
        )
        return LocalTarget(out_dir / "best-vw-command.json")


class VWModel(luigi.Task):
    irt_family = luigi.ChoiceParameter(choices=IRT_FAMILIES)
    irt_type = luigi.ChoiceParameter(choices=IRT_TYPES)
    feature_set = luigi.ChoiceParameter(choices=FEATURE_SETS)

    def requires(self):
        return (
            VWTrainTestData(
                irt_family=self.irt_family, irt_type=self.irt_type, feature_set=self.feature_set,
            ),
            VWHyperTune(
                irt_family=self.irt_family, irt_type=self.irt_type, feature_set=self.feature_set,
            ),
        )

    def run(self):
        out_dir = generate_vw_dir_path(
            irt_family=self.irt_family, irt_type=self.irt_type, feature_set=self.feature_set,
        )
        best_vw_command = read_json(out_dir / "best-vw-command.json")["best_vw_command"]
        shell(best_vw_command.replace("current.model", "best.model"))
        test_args = [
            "vw",
            "--testonly",
            "-d",
            str(out_dir / "test.vw.txt"),
            "-i",
            str(out_dir / "best.model"),
            "-p",
            str(out_dir / "test.pred.txt"),
            "-k",
            "--loss_function=logistic",
            "--link=logistic",
        ]
        shell(" ".join(test_args))

    def output(self):
        out_dir = generate_vw_dir_path(
            irt_family=self.irt_family, irt_type=self.irt_type, feature_set=self.feature_set,
        )
        return (
            LocalTarget(out_dir / "best.model"),
            LocalTarget(out_dir / "test.pred.txt"),
        )


class EvaluateVWModel(luigi.Task):
    irt_family = luigi.ChoiceParameter(choices=IRT_FAMILIES)
    irt_type = luigi.ChoiceParameter(choices=IRT_TYPES)
    feature_set = luigi.ChoiceParameter(choices=FEATURE_SETS)

    def requires(self):
        return VWModel(
            irt_family=self.irt_family, irt_type=self.irt_type, feature_set=self.feature_set,
        )

    def run(self):
        out_dir = generate_vw_dir_path(
            irt_family=self.irt_family, irt_type=self.irt_type, feature_set=self.feature_set,
        )
        eval_vw(
            test_examples_file=out_dir / "test.vw.txt",
            test_pred_file=out_dir / "test.pred.txt",
            report_dir=out_dir,
            name=f"vw-{self.feature_set}",
        )

    def output(self):
        out_dir = generate_vw_dir_path(
            irt_family=self.irt_family, irt_type=self.irt_type, feature_set=self.feature_set,
        )
        return (
            LocalTarget(out_dir / "report.json"),
            LocalTarget(out_dir / "roc.pdf"),
            LocalTarget(out_dir / "precision_recall.pdf"),
        )


class AllVWModels(luigi.WrapperTask):
    def requires(self):
        for f in FEATURE_SETS:
            features = conf["vw"][f]["features"]
            if "irt" in features:
                for irt_type in ["1PL", "2PL", "3PL"]:
                    # Only care about pyro model for now
                    yield EvaluateVWModel(irt_family="pyro", irt_type=irt_type, feature_set=f)
            else:
                yield EvaluateVWModel(feature_set=f, irt_family="no_irt", irt_type="no_irt_pl")


class SquadInTopicFormat(luigi.ExternalTask):
    def output(self):
        return luigi.LocalTarget(conf["squad"]["mallet_v2"])


class RunTopicModel(luigi.Task):
    name = luigi.Parameter()

    def requires(self):
        return SquadInTopicFormat()

    def run(self):
        output_dir = Path(conf["topic"][self.name]["output_dir"])
        output_dir.mkdir(parents=True, exist_ok=True)
        model = TopicModel(
            num_topics=conf["topic"][self.name]["num_topics"],
            input_file=conf["squad"]["mallet_v2"],
            output_dir=output_dir,
        )
        model.train()

    def output(self):
        output_dir = Path(conf["topic"][self.name]["output_dir"])
        return [
            luigi.LocalTarget(output_dir / "parameters.json"),
            luigi.LocalTarget(output_dir / "mallet.topic_keys"),
            luigi.LocalTarget(output_dir / "mallet.topic_distributions"),
            luigi.LocalTarget(output_dir / "mallet.model"),
            luigi.LocalTarget(output_dir / "input_data.mallet"),
            luigi.LocalTarget(output_dir / "mallet.state.gz"),
        ]


class AllTopicModels(luigi.WrapperTask):
    def requires(self):
        for name in conf["topic"].keys():
            yield RunTopicModel(name=name)


class StatisticalTests(luigi.Task):
    resources = {"parallel": 1}
    irt_type = luigi.ChoiceParameter(choices=["1PL", "2PL", "3PL"])
    sampling = luigi.EnumParameter(enum=Sampling)
    percent = luigi.IntParameter()
    fold = luigi.ChoiceParameter(choices=["dev", "test"])

    def requires(self):
        yield LeaderboardSubmissionPredictions(fold=self.fold)

    def run(self):
        pseq(list(StatTest)).map(
            lambda t: run_parallel_test(
                StatTest(t),
                irt_type=self.irt_type,
                sampling=self.sampling,
                percent=self.percent,
                fold=self.fold,
            )
        ).list()

    def output(self):
        output_dir = (
            Path("data/stats")
            / f"fold={self.fold}"
            / f"sampling={self.sampling.value}"  # pylint: disable=no-member
            / f"percent={self.percent}"
        )
        return [luigi.LocalTarget(output_dir / f"{test.value}.json") for test in list(StatTest)]


class AllStatisticalTests(luigi.WrapperTask):
    def requires(self):
        folds = ["dev"]
        if Path(conf["squad"]["submission_predictions"]["test"]).exists():
            folds.append("test")

        for fold in folds:
            if fold == "dev":
                for percent in [5, 10, 25, 50, 75, 100]:
                    for sampling in list(Sampling):
                        yield StatisticalTests(
                            irt_type="3PL", sampling=sampling, percent=percent, fold=fold,
                        )
            elif fold == "test":
                yield StatisticalTests(
                    irt_type="3PL", sampling=Sampling.RANDOM, percent=100, fold=fold
                )


class McNemarStatisticalPower(luigi.Task):
    resources = {"parallel": 1}

    def run(self):
        power.mcnemar("data/power/mcnemar_trials.json")

    def output(self):
        return luigi.LocalTarget("data/power/mcnemar_trials.json")
